### ISSUE TITLE
Over-aggressive pruning inside loops with try/catch

### ISSUE THREAD

**Issue by jbt** [rollup/rollup#2894]
- Rollup Version: 1.14.0
- Operating System (or Browser): Any
- Node Version: 10.16.0

### How Do We Reproduce?

Using the following code as input for rollup, no plugins and all the default options:

```js
export default function mixin(Interface, Classes) {

    for (var _iterator2 = Object.getOwnPropertyNames(Interface)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var name = _step2.value;

      if (Class.hasOwnProperty(name)) continue;
      var desc = Object.getOwnPropertyDescriptor(Interface, name);
      Object.defineProperty(Class, name, desc);
    }

    try {
      for (var _iterator3 = Object.getOwnPropertyNames(Interface.prototype)[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
        var _name = _step3.value;

        if (Class.prototype.hasOwnProperty(_name)) continue;
        var desc = Object.getOwnPropertyDescriptor(Interface.prototype, _name);
        Object.defineProperty(Class.prototype, _name, desc);
      }
    } catch (err) {
    }
}

```

(this is a reduced version of the [mixin util from Slate.js](https://github.com/ianstormtaylor/slate/blob/e7ab7f82791eb50805511eeae0a8d38049b43363/packages/slate/src/utils/mixin.js) after transpilation that can be found in `lib/slate.es.js` of slate on npm)

### Expected Behavior

Output should be pretty much identical to input

### Actual Behavior

The second loop gets reduced down to:

```js
      for (; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
        var desc;
      }
```

`_iterator3` is no longer defined and the loop body is also missing.

This seems to be about as minimal as I can make the test case - removing the try/catch or anything inside either of the loops causes this behaviour to go away.

[REPL link](http://rollupjs.org/repl?version=1.14.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGRlZmF1bHQlMjBmdW5jdGlvbiUyMG1peGluKEludGVyZmFjZSUyQyUyMENsYXNzZXMpJTIwJTdCJTVDbiU1Q24lMjAlMjAlMjAlMjBmb3IlMjAodmFyJTIwX2l0ZXJhdG9yMiUyMCUzRCUyME9iamVjdC5nZXRPd25Qcm9wZXJ0eU5hbWVzKEludGVyZmFjZSklNUJTeW1ib2wuaXRlcmF0b3IlNUQoKSUyQyUyMF9zdGVwMiUzQiUyMCEoX2l0ZXJhdG9yTm9ybWFsQ29tcGxldGlvbjIlMjAlM0QlMjAoX3N0ZXAyJTIwJTNEJTIwX2l0ZXJhdG9yMi5uZXh0KCkpLmRvbmUpJTNCJTIwX2l0ZXJhdG9yTm9ybWFsQ29tcGxldGlvbjIlMjAlM0QlMjB0cnVlKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB2YXIlMjBuYW1lJTIwJTNEJTIwX3N0ZXAyLnZhbHVlJTNCJTVDbiU1Q24lMjAlMjAlMjAlMjAlMjAlMjBpZiUyMChDbGFzcy5oYXNPd25Qcm9wZXJ0eShuYW1lKSklMjBjb250aW51ZSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB2YXIlMjBkZXNjJTIwJTNEJTIwT2JqZWN0LmdldE93blByb3BlcnR5RGVzY3JpcHRvcihJbnRlcmZhY2UlMkMlMjBuYW1lKSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjBPYmplY3QuZGVmaW5lUHJvcGVydHkoQ2xhc3MlMkMlMjBuYW1lJTJDJTIwZGVzYyklM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU1Q24lMjAlMjAlMjAlMjB0cnklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwZm9yJTIwKHZhciUyMF9pdGVyYXRvcjMlMjAlM0QlMjBPYmplY3QuZ2V0T3duUHJvcGVydHlOYW1lcyhJbnRlcmZhY2UucHJvdG90eXBlKSU1QlN5bWJvbC5pdGVyYXRvciU1RCgpJTJDJTIwX3N0ZXAzJTNCJTIwIShfaXRlcmF0b3JOb3JtYWxDb21wbGV0aW9uMyUyMCUzRCUyMChfc3RlcDMlMjAlM0QlMjBfaXRlcmF0b3IzLm5leHQoKSkuZG9uZSklM0IlMjBfaXRlcmF0b3JOb3JtYWxDb21wbGV0aW9uMyUyMCUzRCUyMHRydWUpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHZhciUyMF9uYW1lJTIwJTNEJTIwX3N0ZXAzLnZhbHVlJTNCJTVDbiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBpZiUyMChDbGFzcy5wcm90b3R5cGUuaGFzT3duUHJvcGVydHkoX25hbWUpKSUyMGNvbnRpbnVlJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHZhciUyMGRlc2MlMjAlM0QlMjBPYmplY3QuZ2V0T3duUHJvcGVydHlEZXNjcmlwdG9yKEludGVyZmFjZS5wcm90b3R5cGUlMkMlMjBfbmFtZSklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwT2JqZWN0LmRlZmluZVByb3BlcnR5KENsYXNzLnByb3RvdHlwZSUyQyUyMF9uYW1lJTJDJTIwZGVzYyklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCUyMCUyMCU3RCUyMGNhdGNoJTIwKGVyciklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU3RCU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==)

This seems to have been introduced in 1.14.0, since with 1.13.1 it's perfectly fine ([repl](http://rollupjs.org/repl?version=1.13.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGRlZmF1bHQlMjBmdW5jdGlvbiUyMG1peGluKEludGVyZmFjZSUyQyUyMENsYXNzZXMpJTIwJTdCJTVDbiU1Q24lMjAlMjAlMjAlMjBmb3IlMjAodmFyJTIwX2l0ZXJhdG9yMiUyMCUzRCUyME9iamVjdC5nZXRPd25Qcm9wZXJ0eU5hbWVzKEludGVyZmFjZSklNUJTeW1ib2wuaXRlcmF0b3IlNUQoKSUyQyUyMF9zdGVwMiUzQiUyMCEoX2l0ZXJhdG9yTm9ybWFsQ29tcGxldGlvbjIlMjAlM0QlMjAoX3N0ZXAyJTIwJTNEJTIwX2l0ZXJhdG9yMi5uZXh0KCkpLmRvbmUpJTNCJTIwX2l0ZXJhdG9yTm9ybWFsQ29tcGxldGlvbjIlMjAlM0QlMjB0cnVlKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB2YXIlMjBuYW1lJTIwJTNEJTIwX3N0ZXAyLnZhbHVlJTNCJTVDbiU1Q24lMjAlMjAlMjAlMjAlMjAlMjBpZiUyMChDbGFzcy5oYXNPd25Qcm9wZXJ0eShuYW1lKSklMjBjb250aW51ZSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB2YXIlMjBkZXNjJTIwJTNEJTIwT2JqZWN0LmdldE93blByb3BlcnR5RGVzY3JpcHRvcihJbnRlcmZhY2UlMkMlMjBuYW1lKSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjBPYmplY3QuZGVmaW5lUHJvcGVydHkoQ2xhc3MlMkMlMjBuYW1lJTJDJTIwZGVzYyklM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU1Q24lMjAlMjAlMjAlMjB0cnklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwZm9yJTIwKHZhciUyMF9pdGVyYXRvcjMlMjAlM0QlMjBPYmplY3QuZ2V0T3duUHJvcGVydHlOYW1lcyhJbnRlcmZhY2UucHJvdG90eXBlKSU1QlN5bWJvbC5pdGVyYXRvciU1RCgpJTJDJTIwX3N0ZXAzJTNCJTIwIShfaXRlcmF0b3JOb3JtYWxDb21wbGV0aW9uMyUyMCUzRCUyMChfc3RlcDMlMjAlM0QlMjBfaXRlcmF0b3IzLm5leHQoKSkuZG9uZSklM0IlMjBfaXRlcmF0b3JOb3JtYWxDb21wbGV0aW9uMyUyMCUzRCUyMHRydWUpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHZhciUyMF9uYW1lJTIwJTNEJTIwX3N0ZXAzLnZhbHVlJTNCJTVDbiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBpZiUyMChDbGFzcy5wcm90b3R5cGUuaGFzT3duUHJvcGVydHkoX25hbWUpKSUyMGNvbnRpbnVlJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHZhciUyMGRlc2MlMjAlM0QlMjBPYmplY3QuZ2V0T3duUHJvcGVydHlEZXNjcmlwdG9yKEludGVyZmFjZS5wcm90b3R5cGUlMkMlMjBfbmFtZSklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwT2JqZWN0LmRlZmluZVByb3BlcnR5KENsYXNzLnByb3RvdHlwZSUyQyUyMF9uYW1lJTJDJTIwZGVzYyklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCUyMCUyMCU3RCUyMGNhdGNoJTIwKGVyciklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU3RCU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==))

**Comment by lukastaegert** [comment]
Thanks, I will look into this, this seems strange indeed. Weirdly enough, the issue seems to arise from the variable `desc` that is defined inside the try-statement but accessed outside it. https://rollupjs.org/repl?version=1.14.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnNvbGUubG9nKGRlc2MpJTNCJTVDbiU1Q250cnklMjAlN0IlNUNuJTIwJTIwZm9yJTIwKHZhciUyMGklMjAlM0QlMjAxJTNCJTIwaSUyMCUzQyUyMDIlM0IlMjBpJTJCJTJCKSUyMCU3QiU1Q24lNUN0JTVDdGNvbnNvbGUubG9nKCdub29uZSUyMGNhcmVzJyklM0IlNUNuJTIwJTIwJTIwJTIwdmFyJTIwZGVzYyUyMCUzRCUyMCU3QiU3RCUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEJTIwY2F0Y2glMjAoZXJyKSUyMCU3QiU3RCU1Q24lNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXNtJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

**Comment by colinrcummings** [comment]
I ran into this as well building [this small repo](https://github.com/colinrcummings/react-beautiful-dnd-test-utils).

**Comment by lukastaegert** [comment]
Thanks, I already tracked it down, fix upcoming.

**Comment by lukastaegert** [comment]
Fix at #2898
